drivers/driver.py

```diff
@@ -0,0 +1,169 @@
+
+    if np.isscalar(data):
+        data = np.array([data], data.dtype)
+    buf = data.tobytes()
+    sock.send(buf)
+
+
+HDRLEN = 12  # number of characters of the default message strings
+
+
+def Message(mystr):
+    """Returns a header of standard length HDRLEN."""
+
+    # convert to bytestream since we'll be sending this over a socket
+    return str.ljust(str.upper(mystr), HDRLEN).encode()
+
+
+def dummy_driver(cell, pos):
+    """ Does nothing, but returns properties that can be used by the driver loop."""
+    pot = 0.0
+    force = pos * 0.0  # makes a zero force with same shape as pos
+    vir = cell * 0.0  # makes a zero virial with same shape as cell
+    extras = "nada"
+    return pot, force, vir, extras
+
+
+def driver(unix=False, address="", port=12345, driver=dummy_driver):
+    """Minimal socket client for i-PI."""
+
+    # Opens a socket to i-PI
+    if unix:
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect("/tmp/ipi_" + address)
+    else:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        raise ValueError("We haven't yet implemented the mf")
+
+    f_init = False
+    f_data = False
+    f_verbose = False
+
+    # initializes structure arrays
+    cell = np.zeros((3, 3), float)
+    icell = np.zeros((3, 3), float)
+    pos = np.zeros(0, float)
+
+    # initializes return arrays
+    pot = 0.0
+    force = np.zeros(0, float)
+    vir = np.zeros((3, 3), float)
+    while True:  # ah the infinite loop!
+        header = sock.recv(HDRLEN)
+
+        if header == Message("STATUS"):
+            # responds to a status request
+            if not f_init:
+                sock.sendall(Message("NEEDINIT"))
+            elif f_data:
+                sock.sendall(Message("HAVEDATA"))
+            else:
+                sock.sendall(Message("READY"))
+        elif header == Message("INIT"):
+            # initialization
+            rid = recv_data(sock, np.int32())
+            initlen = recv_data(sock, np.int32())
+            initstr = recv_data(sock, np.chararray(initlen))
+            if f_verbose:
+                print(rid, initstr)
+            f_init = True  # we are initialized now
+        elif header == Message("POSDATA"):
+            # receives structural information
+            cell = recv_data(sock, cell)
+            icell = recv_data(
+                sock, icell
+            )  # inverse of the cell. mostly useless legacy stuff
+            nat = recv_data(sock, np.int32())
+            if len(pos) == 0:
+                # shapes up the position array
+                pos.resize((nat, 3))
+                force.resize((nat, 3))
+            else:
+                if len(pos) != nat:
+                    raise RuntimeError("Atom number changed during i-PI run")
+            pos = recv_data(sock, pos)
+
+            ##### THIS IS THE TIME TO DO SOMETHING WITH THE POSITIONS!
+            pot, force, vir, extras = driver(cell, pos)
```

Comment:
shouldn't this be dummy_driver instead of driver?

Reply:
ah crap this is confusing  I agree